collections/src/main/scala/strawman/collection/mutable/WrappedArray.scala

```diff
@@ -0,0 +1,310 @@
+  *  underlying `Array`, therefore it is not growable or shrinkable.
+  *
+  *  @tparam T    type of the elements in this wrapped array.
+  *
+  *  @author  Martin Odersky, Stephane Micheloud
+  *  @version 1.0
+  *  @since 2.8
+  *  @define Coll `WrappedArray`
+  *  @define coll wrapped array
+  *  @define orderDependent
+  *  @define orderDependentFold
+  *  @define mayNotTerminateInf
+  *  @define willNotTerminateInf
+  */
+abstract class WrappedArray[T]
+  extends AbstractSeq[T]
+    with IndexedSeq[T]
+    with IndexedSeqOps[T, WrappedArray, WrappedArray[T]]
+    with ArrayLike[T]
+    with IndexedOptimizedSeq[T]
+    with StrictOptimizedSeqOps[T, WrappedArray, WrappedArray[T]]
+    with Serializable {
+
+  def iterableFactory: strawman.collection.SeqFactory[WrappedArray] = WrappedArray
+
+  protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[T]): WrappedArray[T] = {
+    val b = new WrappedArrayBuilder(elemTag)
+    val s = coll.knownSize
+    if(s > 0) b.sizeHint(s)
+    b ++= coll
+    b.result()
+  }
+  protected[this] def newSpecificBuilder(): Builder[T, WrappedArray[T]] = new WrappedArrayBuilder(elemTag)
+
+  /** The tag of the element type */
+  def elemTag: ClassTag[T]
+
+  /** Update element at given index */
+  def update(index: Int, elem: T): Unit
+
+  /** The underlying array */
+  def array: Array[T]
+
+  override def toArray[U >: T : ClassTag]: Array[U] = {
+    val thatElementClass = implicitly[ClassTag[U]].runtimeClass
+    if (array.getClass.getComponentType eq thatElementClass)
+      array.asInstanceOf[Array[U]]
+    else
+      super.toArray[U]
+  }
+
+  override def className = "WrappedArray"
+
+  /** Clones this object, including the underlying Array. */
+  override def clone(): WrappedArray[T] = WrappedArray.make(array.clone())
+}
+
+/** A companion object used to create instances of `WrappedArray`.
+  */
+object WrappedArray extends StrictOptimizedSeqFactory[WrappedArray] {
+  // This is reused for all calls to empty.
+  private val EmptyWrappedArray  = new ofRef[AnyRef](new Array[AnyRef](0))
+  def empty[T]: WrappedArray[T] = EmptyWrappedArray.asInstanceOf[WrappedArray[T]]
+
+  def from[A](it: strawman.collection.IterableOnce[A]): WrappedArray[A] = {
+    val n = it.knownSize
+    if (n > -1) {
+      val elements = scala.Array.ofDim[Any](n)
+      val iterator = it.iterator()
+      var i = 0
+      while (i < n) {
+        ScalaRunTime.array_update(elements, i, iterator.next())
+        i = i + 1
+      }
+      make(elements)
+    } else fromArrayBuffer(ArrayBuffer.from(it))
+  }
+
+  def fromArrayBuffer[A](arr: ArrayBuffer[A]): WrappedArray[A] =
+    make[A](arr.asInstanceOf[ArrayBuffer[Any]].toArray)
+
+  def newBuilder[A](): Builder[A, WrappedArray[A]] =
+    ArrayBuffer.newBuilder[A]().mapResult(fromArrayBuffer)
+
+  def newBuilder[A](implicit t: ClassTag[A]): Builder[A, WrappedArray[A]] = new WrappedArrayBuilder[A](t)
```

Comment:
Why do we need `WrappedArrayBuilder`? Can’t we use `ArrayBuilder.mapResult(fromArray)`?

Reply:
I don't see any reason why we couldn't.